### PR TITLE
Mudrod 168: Full Ingest 03/2017 logs fails

### PR DESCRIPTION
**Pull request by quintinali**
check whether UserHistoryMatrix.csv or ClickstreamMatrix.csv exists or is empty

### CODE REVIEW

File: core/src/main/java/gov/nasa/jpl/mudrod/utils/MatrixUtil.java [hunk 1]
```diff
@@ -371,7 +386,15 @@ public Boolean call(Tuple2<String, Long> v1) throws Exception {
         String[] fields = t._1.split(",");
         String word = fields[0];
         int fieldsize = fields.length;
-        String[] numfields = Arrays.copyOfRange(fields, 1, fieldsize - 1);
+        int nStart = 1;
+        int nEnd = fieldsize - 1;
+        if (fieldsize < 2) {
+          System.out.println("***************output null***************");
```
Comment (fgreg) [inline core/src/main/java/gov/nasa/jpl/mudrod/utils/MatrixUtil.java]: Delete this or make it a useful logging statement.

**Review by fgreg — CHANGES_REQUESTED** [review]
Can you explain what you did exactly? It seems like this is just guarding against the case when clickstream/userhistory is empty, which is good, but isn't the root of the problem the fact that clickstream/userhistory is empty? Does this address that problem?

File: core/src/main/java/gov/nasa/jpl/mudrod/utils/MatrixUtil.java (hunk 1 above)
Comment (fgreg) [reply core/src/main/java/gov/nasa/jpl/mudrod/utils/MatrixUtil.java]: I'm talking about just the` System.out.println("***************output null***************");` statement.

**Comment by quintinali** [comment]
@fgreg It is reasonable sometimes user history and clickstream matrix is empty

(1) user history matrix is what users searched in selected time. Each row is a user and each column is a query. A query should be searched multiple times which is larger than a predefined threshold or it will be discarded. 
(2) Clickstream data record which dataset users click or download after search a query. If users didn't click or download dataset after search a query, the matrix will be empty.
(3) In order to get more accurate sessions,  a session is discarded if it doesn't contain view or download behaviours. For February and March 2017, I can't find FTP logs ( record download behaviors). I think this is the main reason. 

So I decided to solve bugs related to the case when clickstream/userhistory is empty at first.

**Comment by fgreg** [comment]
@quintinali ok, thanks for explaining. I'll test more recent logs and make sure it works. Do please remove that println statement though.

File: core/src/main/java/gov/nasa/jpl/mudrod/weblog/process/ClickStreamAnalyzer.java [hunk 2]
```diff
@@ -49,18 +50,22 @@ public Object execute() {
...
-      svd.getSVDMatrix(props.getProperty("clickstreamMatrix"), Integer.parseInt(props.getProperty("clickstreamSVDDimension")), props.getProperty("clickstreamSVDMatrix_tmp"));
-      List<LinkageTriple> tripleList = svd.calTermSimfromMatrix(props.getProperty("clickstreamSVDMatrix_tmp"));
-      svd.saveToES(tripleList, props.getProperty("indexName"), props.getProperty("clickStreamLinkageType"));
+      String clickstream_matrixFile = props.getProperty("clickstreamMatrix");
+      File f = new File(clickstream_matrixFile);
+      if (f.exists()) {
+        SVDAnalyzer svd = new SVDAnalyzer(props, es, spark);
+        svd.getSVDMatrix(props.getProperty("clickstreamMatrix"), Integer.parseInt(props.getProperty("clickstreamSVDDimension")), props.getProperty("clickstreamSVDMatrix_tmp"));
+        List<LinkageTriple> tripleList = svd.calTermSimfromMatrix(props.getProperty("clickstreamSVDMatrix_tmp"));
+        svd.saveToES(tripleList, props.getProperty("indexName"), props.getProperty("clickStreamLinkageType"));
```
Comment (lewismc) [inline core/src/main/java/gov/nasa/jpl/mudrod/weblog/process/ClickStreamAnalyzer.java:59]: @quintinali please use MudrodConstants for all of these property lookup's. Thank you

**Review by lewismc — CHANGES_REQUESTED** [review]
Please consider comments @quintinali thank you

**Comment by lewismc** [comment]
Hi @quintinali this seems to have fixed the bug. The only issue I get now is as follows
```
2017-05-16 20:56:38,203 INFO  process.MetadataAnalyzer (MetadataAnalyzer.java:execute(66)) - *****************Metadata Analyzer starts******************
org.apache.hadoop.mapred.InvalidInputException: Input path does not exist: file:/usr/local/mudrod_resources/weekly_logs/2016/11/w1/MetadataMatrix.csv
	at org.apache.hadoop.mapred.FileInputFormat.listStatus(FileInputFormat.java:251)
	at org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:270)
	at org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:202)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:252)
	at org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:250)
	at scala.Option.getOrElse(Option.scala:121)
	at org.apache.spark.rdd.RDD.partitions(RDD.scala:250)
	at org.apache.spark.rdd.ZippedWithIndexRDD.<init>(ZippedWithIndexRDD.scala:44)
	at org.apache.spark.rdd.RDD$$anonfun$zipWithIndex$1.apply(RDD.scala:1293)
	at org.apache.spark.rdd.RDD$$anonfun$zipWithIndex$1.apply(RDD.scala:1293)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
	at org.apache.spark.rdd.RDD.zipWithIndex(RDD.scala:1292)
	at org.apache.spark.api.java.JavaRDDLike$class.zipWithIndex(JavaRDDLike.scala:342)
	at org.apache.spark.api.java.AbstractJavaRDDLike.zipWithIndex(JavaRDDLike.scala:45)
	at gov.nasa.jpl.mudrod.utils.MatrixUtil.loadVectorFromCSV(MatrixUtil.java:359)
	at gov.nasa.jpl.mudrod.semantics.SVDAnalyzer.getSVDMatrix(SVDAnalyzer.java:59)
	at gov.nasa.jpl.mudrod.metadata.process.MetadataAnalyzer.execute(MetadataAnalyzer.java:74)
	at gov.nasa.jpl.mudrod.discoveryengine.MetadataDiscoveryEngine.process(MetadataDiscoveryEngine.java:67)
	at gov.nasa.jpl.mudrod.main.MudrodEngine.startFullIngest(MudrodEngine.java:261)
	at gov.nasa.jpl.mudrod.main.MudrodEngine.main(MudrodEngine.java:401)
2017-05-16 20:56:38,224 INFO  process.MetadataAnalyzer (MetadataAnalyzer.java:execute(85)) - *****************Metadata Analyzer ends******************Took 0s
```

**Comment by lewismc** [comment]
I am +1 to merge this pull request into master branch and we can work towards improving logging and the error handling in subsequent pull requests.

**Comment by quintinali** [comment]
@Yongyao Could you please take a look at the issue Lewis mentioned above. I think it is related to metadata harvest and analysis.

**Comment by Yongyao** [comment]
@lewismc Can you elaborate on what data and command you were using?

**Comment by lewismc** [comment]
Yes sorry
Data is from [here](ftp://podaac.jpl.nasa.gov/misc/outgoing/cjf/mudrod/2017/04/)
Command is as follows
```
./core/target/appassembler/bin/mudrod-engine -f -dataDir /usr/local/mudrod_resources/weekly_logs/2017/04/w1/
```
I took WWW, WWWssl and FTP logs for week 1 and put them into ```/usr/local/mudrod_resources/weekly_logs/2017/04/w1/```

**Pull request merged**